com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl

```diff
@@ -878,23 +899,32 @@ half4 UniversalFragmentPBR(InputData inputData, SurfaceData surfaceData)
     half3 color = GlobalIllumination(brdfData, brdfDataClearCoat, surfaceData.clearCoatMask,
                                      inputData.bakedGI, surfaceData.occlusion,
                                      inputData.normalWS, inputData.viewDirectionWS);
-    color += LightingPhysicallyBased(brdfData, brdfDataClearCoat,
-                                     mainLight,
-                                     inputData.normalWS, inputData.viewDirectionWS,
-                                     surfaceData.clearCoatMask, specularHighlightsOff);
+
+    if (IsMatchingLightLayer(mainLight.lightLayers, meshRenderingLayers))
+    {
+        color += LightingPhysicallyBased(brdfData, brdfDataClearCoat,
+                                         mainLight,
+                                         inputData.normalWS, inputData.viewDirectionWS,
+                                         surfaceData.clearCoatMask, specularHighlightsOff);
+    }
 
 #ifdef _ADDITIONAL_LIGHTS
     uint pixelLightCount = GetAdditionalLightsCount();
     LIGHT_LOOP_BEGIN(pixelLightCount)
         Light light = GetAdditionalLight(lightIndex, inputData.positionWS, shadowMask);
-        #if defined(_SCREEN_SPACE_OCCLUSION)
-            light.color *= aoFactor.directAmbientOcclusion;
-        #endif
-        color += LightingPhysicallyBased(brdfData, brdfDataClearCoat,
-                                         light,
-                                         inputData.normalWS, inputData.viewDirectionWS,
-                                         surfaceData.clearCoatMask, specularHighlightsOff);
+
+        if (IsMatchingLightLayer(light.lightLayers, meshRenderingLayers))
```

Comment:
Does our compiler create a branch here if light layers are enabled?

Reply:
Don't know for mobile or urp but in HDRP yes this code create a dynamic branch. On AMD GCN architecture it is a scalar branch so it is efficeint, but don't know for low level platform

Reply:
In next commit https://github.com/Unity-Technologies/Graphics/pull/3677/commits/bf099f9f22df2d385cb535f2b6d69df3ce1ec439, I adjusted the code so that IsMatchingLightLayer() is compiled out when light layers is not enabled. I confirmed the code is compiled out on Nintendo Switch since we can see actual assembly, I would expect the same for desktop and consoles. I'm not sure about the fate on mobiles.

Reply:
@kaychang-unity so is the compiler smart to strip out code when using DEFAULT_LIGHT_LAYERS?

Reply:
It is smart enough on Switch as the referenced variables (light.lightlayers, meshRenderingLayers) becomes values known at compile time. I think it will be the same on all platforms.